handsontable/src/plugins/nestedHeaders/nestedHeaders.js

```diff
@@ -518,37 +519,48 @@ export class NestedHeaders extends BasePlugin {
    * @param {Function} clipboardData.getData Gets copied data stored as array of arrays.
    * @param {Function} clipboardData.getMetaInfo Gets grid settings for copied data.
+   * @param {Function} clipboardData.setMetaInfo Sets grid settings for copied data.
    * @param {Function} clipboardData.getRanges Returns ranges related to copied part of Handsontable.
    */
   onBeforeCopy(clipboardData) {
-    const copyableRanges = clipboardData.getRanges();
+    const { rows, columns } = normalizeRanges(clipboardData.getRanges());
+    const headers = rows.filter(row => row < 0);
+    const nestedHeaders = [];
 
-    for (let rangeIndex = 0; rangeIndex < copyableRanges.length; rangeIndex += 1) {
-      const { startRow, startCol, endRow, endCol } = copyableRanges[rangeIndex];
-      const columnsCount = startCol - endCol + 1;
+    headers.forEach((header) => {
+      const headersForLevel = [];
 
-      // do not process dataset ranges and column headers where only one column is copied
-      if (startRow >= 0 || columnsCount === 1) {
-        break;
-      }
+      columns.forEach((column, index) => {
+        const { isPlaceholder } = this.getHeaderSettings(header, column);
+        const columnsToEnd = columns.slice(index).length;
+        const headerTreeNodeData = this.getStateManager().getHeaderTreeNodeData(header, column);
+
+        if (index === 0 && isPlaceholder === true) {
+          const { label, origColspan: colspan, columnIndex } = headerTreeNodeData;
+          const columnFromStart = column - columnIndex;
+          const reducedColspan = Math.min(colspan - columnFromStart, columnsToEnd);
 
-      for (let column = startCol; column <= endCol; column += 1) {
-        for (let row = startRow; row <= endRow; row += 1) {
-          const zeroBasedColumnIndex = column - startCol;
+          if (reducedColspan > 1) {
+            headersForLevel.push({ label, colspan: reducedColspan });
 
-          if (zeroBasedColumnIndex === 0) {
-            continue; // eslint-disable-line no-continue
+          } else {
+            headersForLevel.push(label);
           }
 
-          const isRoot = this.#stateManager.getHeaderTreeNodeData(row, column)?.isRoot;
-          const collapsible = this.#stateManager.getHeaderTreeNodeData(row, column)?.collapsible;
+        } else if (isPlaceholder === false) {
+          const { label, origColspan: colspan, columnIndex } = headerTreeNodeData;
+          const columnFromStart = column - columnIndex;
+          const reducedColspan = Math.min(colspan - columnFromStart, columnsToEnd);
 
-          if (collapsible === true && isRoot === false) {
-            clipboardData.setCellAt(row, zeroBasedColumnIndex, '');
-          }
+          headersForLevel.push(reducedColspan > 1 ? { label, colspan: reducedColspan } : label);
         }
-      }
-    }
+      });
+
+      nestedHeaders.push(headersForLevel);
+    });
+
+    clipboardData.setMetaInfo('colHeaders', undefined);
```

Comment:
```suggestion
    clipboardData.setMetaInfo('colHeaders', null);
    // or, even better
    clipboardData.removeMetaInfo('colHeaders');
```
What if the `setMetaInfo` method in the future has a second optional argument? Then passing  `undefined` value would cause a bug in that case.